spokestack/asr/google/speech_recognizer.py

```diff
@@ -0,0 +1,104 @@
+"""
+This module contains the google asr speech recognizer
+"""
+import logging
+from queue import Queue
+from threading import Thread
+from typing import Any
+
+import numpy as np  # type: ignore
+from google.cloud import speech  # type: ignore
+from google.oauth2 import service_account  # type: ignore
+
+from spokestack.context import SpeechContext
+
+
+_LOG = logging.getLogger(__name__)
+
+
+class GoogleSpeechRecognizer:
+    """ Transforms speech into text using Google's ASR.
+
+     Args:
+         cred_path (str): path to google credentials
+         language (str): language of input audio
+         sample_rate (int): sample rate of the input audio (Hz)
+         **kwargs (optional): additional keyword arguments
+     """
+
+    def __init__(
+        self, cred_path, language="en-US", sample_rate=16000, **kwargs,
+    ) -> None:
+
+        credentials = service_account.Credentials.from_service_account_file(cred_path)
+        self._client = speech.SpeechClient(credentials=credentials, **kwargs)
+        self._config = speech.StreamingRecognitionConfig(
+            config=speech.RecognitionConfig(
+                encoding=speech.RecognitionConfig.AudioEncoding.LINEAR16,
+                sample_rate_hertz=sample_rate,
+                language_code=language,
+                enable_automatic_punctuation=True,
+            ),
+            interim_results=False,
+        )
+        self._sample_rate = sample_rate
+        self._language = language
+        self._queue: Queue = Queue()
+        self._thread: Any = None
+
+    def __call__(self, context: SpeechContext, frame: np.ndarray) -> None:
+        """ Main entry point.
+
+        Args:
+            context (SpeechContext): current state of the speech pipeline
+            frame (np.ndarray): numpy array of PCM-16 audio.
+
+        Returns:
+
+        """
+        if self._thread is None and context.is_active:
+            self._begin(context)
+        if self._thread is not None and not context.is_active:
+            self._commit()
+        if context.is_active:
+            self._send(frame)
+
+    def _begin(self, context) -> None:
+        self._thread = Thread(target=self._receive, args=(context,),)
+        self._thread.start()
+
+    def _receive(self, context):
+
+        for response in self._client.streaming_recognize(self._config, self._drain()):
+            for result in response.results[:1]:
+                for alternative in result.alternatives[:1]:
+                    context.transcript = alternative.transcript
+                    context.confidence = alternative.confidence
+
+                if result.is_final:
```

Comment:
we should also support partial result events - see how this is done in the spokestack asr client